File: cs/Minsk.Tests/CodeAnalysis/EvaluatorTests.cs
﻿using Minsk.CodeAnalysis;
using Minsk.CodeAnalysis.Syntax;
using Minsk.Tests.CodeAnalysis.Text;
using Xunit;

namespace Minsk.Tests.CodeAnalysis;

public sealed class EvaluatorTests
{
    [Theory]
    [InlineData("1", 1)]
    [InlineData("+1", 1)]
    [InlineData("-1", -1)]
    [InlineData("14 + 12", 26)]
    [InlineData("12 - 3", 9)]
    [InlineData("4 * 2", 8)]
    [InlineData("9 / 3", 3)]
    [InlineData("(10)", 10)]
    [InlineData("12 == 3", false)]
    [InlineData("3 == 3", true)]
    [InlineData("12 != 3", true)]
    [InlineData("3 != 3", false)]
    [InlineData("3 < 4", true)]
    [InlineData("5 < 4", false)]
    [InlineData("3 <= 4", true)]
    [InlineData("4 <= 4", true)]
    [InlineData("5 <= 4", false)]
    [InlineData("3 > 4", false)]
    [InlineData("5 > 4", true)]
    [InlineData("3 >= 4", false)]
    [InlineData("4 >= 4", true)]
    [InlineData("5 >= 4", true)]
    [InlineData("1 | 2", 3)]
    [InlineData("1 | 0", 1)]
    [InlineData("1 & 3", 1)]
    [InlineData("1 & 0", 0)]
    [InlineData("1 ^ 2", 3)]
    [InlineData("1 ^ 3", 2)]
    [InlineData("~1", -2)]
    [InlineData("false == false", true)]
    [InlineData("true == false", false)]
    [InlineData("false != false", false)]
    [InlineData("true != false", true)]
    [InlineData("true && true", true)]
    [InlineData("true && false", false)]
    [InlineData("false || false", false)]
    [InlineData("true || false", true)]
    [InlineData("false | false", false)]
    [InlineData("true | false", true)]
    [InlineData("false & false", false)]
    [InlineData("true & false", false)]
    [InlineData("false ^ false", false)]
    [InlineData("true ^ false", true)]
    [InlineData("true ^ true", false)]
    [InlineData("true", true)]
    [InlineData("false", false)]
    [InlineData("!true", false)]
    [InlineData("!false", true)]
    [InlineData("{ var a = 0 (a = 10) * a }", 100)]
    [InlineData("{ var a = 0 if a == 0 a = 10 a }", 10)]
    [InlineData("{ var a = 0 if a == 5 a = 10 a }", 0)]
    [InlineData("{ var a = 0 if a == 0 a = 10 else a = 20 a }", 10)]
    [InlineData("{ var a = 0 if a == 5 a = 10 else a = 20 a }", 20)]
    [InlineData("{ var i = 10 var result = 0 while i > 0 { result = result + i i = i - 1 } result }", 55)]
    [InlineData("{ var result = 0 for i = 1 to 10 { result = result + i } result }", 55)]
    public void EvaluatesCorrectValue(string text, object expected)
    {
        AssertValue(text, expected);
    }

    [Fact]
    public void VariableDeclarationReportsRedeclaration()
    {
        const string text = @"
            {
                var x = 10
                var y = 100
                {
                    var x = 10
                }
                var [x] = 5
            }
        ";
        const string diagnostics = @"
            Name 'x' is already declared in this scope
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void NameExpressionReportsUndefined()
    {
        const string text = @"
            [x] * 10
        ";
        const string diagnostics = @"
            Undefined name 'x'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void AssignmentExpressionReportsUndefined()
    {
        const string text = @"
            [x] = 10
        ";
        const string diagnostics = @"
            Undefined name 'x'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void AssignmentExpressionReportsReadOnly()
    {
        const string text = @"
            {
                let x = 10
                x [=] 20
            }
        ";
        const string diagnostics = @"
            Cannot assign to read-only variable 'x'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void AssignmentExpressionReportsCannotConvert()
    {
        const string text = @"
            {
                var x = 10
                x [=] true
            }
        ";
        const string diagnostics = @"
            Cannot convert type 'System.Boolean' to 'System.Int32'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void UnaryOperatorReportsUndefined()
    {
        const string text = @"
            [+]true
        ";
        const string diagnostics = @"
            Unary operator '+' is not defined for type 'System.Boolean'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void BinaryOperatorReportsUndefined()
    {
        const string text = @"
            true [+] true
        ";
        const string diagnostics = @"
            Binary operator '+' is not defined for types 'System.Boolean' and 'System.Boolean'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void IfStatementReportsCannotConvert()
    {
        const string text = @"
            {
                var x = 0
                if [10]
                    x = 10
            }
        ";

        const string diagnostics = @"
            Cannot convert type 'System.Int32' to 'System.Boolean'
        ";

        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void BlockStatementNoInfiniteLoop()
    {
        const string text = @"
            {
            [)][]
        ";
        const string diagnostics = @"
            Expected next token to be IdentifierToken, got CloseParenthesisToken instead
            Expected next token to be CloseBraceToken, got EndOfFileToken instead
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void NameExpressionReportsNoErrorForInsertedToken()
    {
        const string text = "[]";
        const string diagnostics = @"
            Expected next token to be IdentifierToken, got EndOfFileToken instead
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void WhileStatementReportsCannotConvert()
    {
        const string text = @"
            {
                var x = 0
                while [10]
                    x = 10
            }
        ";

        const string diagnostics = @"
            Cannot convert type 'System.Int32' to 'System.Boolean'
        ";

        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void ForStatementReportsCannotConvertLowerBound()
    {
        const string text = @"
            {
                var result = 0
                for i = [false] to 10
                    result = result + i
            }
        ";
        const string diagnostics = @"
            Cannot convert type 'System.Boolean' to 'System.Int32'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    [Fact]
    public void ForStatementReportsCannotConvertUpperBound()
    {
        const string text = @"
            {
                var result = 0
                for i = 1 to [true]
                    result = result + i
            }
        ";
        const string diagnostics = @"
            Cannot convert type 'System.Boolean' to 'System.Int32'
        ";
        AssertDiagnostics(text, diagnostics);
    }

    private static void AssertValue(string text, object expected)
    {
        var syntaxTree = SyntaxTree.Parse(text);
        var compilation = new Compilation(syntaxTree);
        var variables = new Dictionary<VariableSymbol, object>();
        var result = compilation.Evaluate(variables);

        Assert.Empty(result.Diagnostics);
        Assert.Equal(expected, result.Value);
    }

    private static void AssertDiagnostics(string text, string diagnosticText)
    {
        var annotatedText = AnnotatedText.Parse(text);
        var syntaxTree = SyntaxTree.Parse(annotatedText.Text);
        var compilation = new Compilation(syntaxTree);
        var variables = new Dictionary<VariableSymbol, object>();
        var result = compilation.Evaluate(variables);
        var diagnostics = AnnotatedText.UnindentLines(diagnosticText);

        Assert.Equal(diagnostics.Count, annotatedText.Spans.Length);
        Assert.Equal(diagnostics.Count, result.Diagnostics.Length);

        for (var i = 0; i < diagnostics.Count; i++)
        {
            var expectedMessage = diagnostics[i];
            var actualMessage = result.Diagnostics[i].Message;

            Assert.Equal(expectedMessage, actualMessage);

            var expectedSpan = annotatedText.Spans[i];
            var actualSpan = result.Diagnostics[i].Span;

            Assert.Equal(expectedSpan, actualSpan);
        }
    }
}

File: zig/minsk_tests/src/code_analysis/syntax/evaluator_tests.zig
const std = @import("std");
const t = @import("framework");
const Object = @import("minsk_runtime").Object;
const Compilation = @import("minsk").code_analysis.Compilation;
const SyntaxTree = @import("minsk").code_analysis.syntax.SyntaxTree;
const VariableSymbol = @import("minsk").code_analysis.VariableSymbol;
const AnnotatedText = @import("AnnotatedText.zig");
const camelToDisplay = @import("../../camel_to_display.zig").camelToDisplay;

const EvaluatorTest = struct {
    text: []const u8,
    expected_value: Object,

    pub fn init(text: []const u8, expected_value: Object) @This() {
        return .{ .text = text, .expected_value = expected_value };
    }
};

fn correctEvaluation(state: *t.TestState, ctx: EvaluatorTest, out_msg: *[]const u8) t.TestExit!void {
    try assertHasValue(state, ctx.text, ctx.expected_value, out_msg);
}

fn assertHasValue(state: *t.TestState, text: []const u8, expected_value: Object, out_msg: *[]const u8) t.TestExit!void {
    const syntax_tree = SyntaxTree.parse(t.allocator, text) catch unreachable;
    var compilation = Compilation.init(t.allocator, syntax_tree) catch unreachable;
    defer compilation.deinit(.with_parents);
    var variables = VariableSymbol.Map.init(t.allocator);
    defer variables.deinit();
    const actual_result = compilation.evaluate(&variables) catch unreachable;
    defer actual_result.deinit(t.allocator);

    switch (actual_result) {
        .success => |value| try t.assert(
            state,
            value.eq(expected_value),
            "value mismatch ({} != {})",
            .{ value, expected_value },
            out_msg,
        ),
        .failure => |diagnostics| {
            const diagnostic_strings = t.allocator.alloc([]const u8, diagnostics.len) catch unreachable;
            for (diagnostics, diagnostic_strings) |d, *s| {
                s.* = std.fmt.allocPrint(t.allocator, "{s}", .{d.message}) catch unreachable;
            }
            defer {
                for (diagnostic_strings) |s| {
                    t.allocator.free(s);
                }
                t.allocator.free(diagnostic_strings);
            }
            const msgs = std.mem.join(t.allocator, "\n", diagnostic_strings) catch unreachable;
            defer t.allocator.free(msgs);
            try t.assert(
                state,
                false,
                "failed with diagnostics:\n{s}\n",
                .{msgs},
                out_msg,
            );
        },
    }
}

fn assertHasDiagnostics(
    state: *t.TestState,
    comptime text: []const u8,
    comptime diagnostic_text: []const u8,
    out_msg: *[]const u8,
) t.TestExit!void {
    const annotated_text = comptime AnnotatedText.parse(text);
    const syntax_tree = SyntaxTree.parse(t.allocator, annotated_text.text) catch unreachable;
    const compilation = Compilation.init(t.allocator, syntax_tree) catch unreachable;
    defer compilation.deinit(.with_parents);
    var variables = VariableSymbol.Map.init(t.allocator);
    defer variables.deinit();
    const result = compilation.evaluate(&variables) catch unreachable;
    defer result.deinit(t.allocator);

    const diagnostics = comptime AnnotatedText.unindentLines(diagnostic_text);
    if (annotated_text.spans.len != diagnostics.len) {
        @compileError(std.fmt.comptimePrint(
            "must mark as many spans as there are diagnostics ({d} != {d})",
            .{ annotated_text.spans.len, diagnostics.len },
        ));
    }

    const actual_diagnostics = switch (result) {
        .success => |val| return t.assert(state, false, "expected failure, got {}", .{val}, out_msg),
        .failure => |d| d,
    };

    if (actual_diagnostics.len != diagnostics.len) {
        const diagnostic_strings = blk: {
            var strings = std.ArrayList([]const u8).init(t.allocator);
            defer {
                for (strings.items) |s|
                    t.allocator.free(s);
                strings.deinit();
            }
            for (actual_diagnostics) |ad| {
                strings.append(
                    std.fmt.allocPrint(t.allocator, "{d}: {s}", .{ ad.span.start, ad.message }) catch unreachable,
                ) catch unreachable;
            }
            break :blk std.mem.join(t.allocator, "\n", strings.items) catch unreachable;
        };
        defer t.allocator.free(diagnostic_strings);
        try t.assert(
            state,
            false,
            "actual diagnostics do not match:\n{s}",
            .{diagnostic_strings},
            out_msg,
        );
    }

    for (annotated_text.spans, diagnostics, actual_diagnostics) |
        expected_span,
        expected_message,
        actual_diagnostic,
    | {
        try t.assert(
            state,
            std.mem.eql(u8, expected_message, actual_diagnostic.message),
            \\diagnostic mismatch
            \\expected: {s}
            \\actually: {s}
        ,
            .{ expected_message, actual_diagnostic.message },
            out_msg,
        );

        try t.assert(
            state,
            std.meta.eql(expected_span, actual_diagnostic.span),
            "expected span {}, got {}",
            .{ expected_span, actual_diagnostic.span },
            out_msg,
        );
    }
}

const one_shot_tests = struct {
    pub fn variableDeclarationReportsRedeclaration(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\  var x = 10
            \\  var y = 100
            \\  {
            \\    var x = 10
            \\  }
            \\  var [x] = 5
            \\}
        ;
        const diagnostics =
            \\Variable 'x' has already been declared.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn blockStatementNoInfiniteLoop(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\[)][]
        ;
        const diagnostics =
            \\Unexpected token <CloseParenthesisToken>, expected <IdentifierToken>.
            \\Unexpected token <EndOfFileToken>, expected <CloseBraceToken>.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn nameExpressionNoErrorForInsertedToken(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text = "[]";
        const diagnostics =
            \\Unexpected token <EndOfFileToken>, expected <IdentifierToken>.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn nameExpressionReportsUndefined(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\[x] * 10
        ;
        const diagnostics =
            \\Variable 'x' doesn't exist.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn assignmentExpressionReportsUndefined(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\[x] = 10
        ;
        const diagnostics =
            \\Variable 'x' doesn't exist.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn assignmentReportsCannotAssign(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\  let x = 10
            \\  x [=] 0
            \\}
        ;
        const diagnostics =
            \\Variable 'x' is read-only and cannot be assigned to.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn assignmentReportsCannotConvert(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\  var x = 10
            \\  x = [true]
            \\}
        ;
        const diagnostics =
            \\Cannot convert type Boolean to type Integer.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn ifStatementReportsCannotConvert(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\  if [10]
            \\    var x = true
            \\}
        ;
        const diagnostics =
            \\Cannot convert type Integer to type Boolean.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn whileStatementReportsCannotConvert(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\  while [10]
            \\    var x = true
            \\}
        ;
        const diagnostics =
            \\Cannot convert type Integer to type Boolean.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn forStatementReportsCannotConvertLowerBound(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\  for i = [true] to 10
            \\    var x = true
            \\}
        ;
        const diagnostics =
            \\Cannot convert type Boolean to type Integer.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn forStatementReportsCannotConvertUpperBound(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\{
            \\  for i = 1 to [false]
            \\    var x = true
            \\}
        ;
        const diagnostics =
            \\Cannot convert type Boolean to type Integer.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn unaryReportsUndefined(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\[+]true
        ;
        const diagnostics =
            \\Unary operator '+' is not defined for type Boolean.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }

    pub fn binaryReportsUndefined(state: *t.TestState, _: void, out_msg: *[]const u8) t.TestExit!void {
        const text =
            \\10 [&&] false
        ;
        const diagnostics =
            \\Binary operator '&&' is not defined for types Integer and Boolean.
        ;
        try assertHasDiagnostics(state, text, diagnostics, out_msg);
    }
};

pub fn evaluatorTestSuite(state: *t.TestState) void {
    const et = EvaluatorTest.init;
    for ([_]EvaluatorTest{
        et("1", .{ .integer = 1 }),
        et("+1", .{ .integer = 1 }),
        et("-1", .{ .integer = -1 }),
        et("~1", .{ .integer = -2 }),
        et("1 + 2", .{ .integer = 3 }),
        et("1 - 2", .{ .integer = -1 }),
        et("3 * 3", .{ .integer = 9 }),
        et("9 / 3", .{ .integer = 3 }),
        et("(10)", .{ .integer = 10 }),
        et("12 ==  3", .{ .boolean = false }),
        et(" 3 ==  3", .{ .boolean = true }),
        et("12 !=  3", .{ .boolean = true }),
        et(" 3 !=  3", .{ .boolean = false }),
        et("3 <  4", .{ .boolean = true }),
        et("5 <  4", .{ .boolean = false }),
        et("4 <= 4", .{ .boolean = true }),
        et("5 <= 4", .{ .boolean = false }),
        et("4 >  3", .{ .boolean = true }),
        et("4 >  5", .{ .boolean = false }),
        et("4 >= 4", .{ .boolean = true }),
        et("4 >= 5", .{ .boolean = false }),
        et("1 | 2", .{ .integer = 3 }),
        et("1 | 0", .{ .integer = 1 }),
        et("1 & 2", .{ .integer = 0 }),
        et("2 & 3", .{ .integer = 2 }),
        et("2 ^ 3", .{ .integer = 1 }),
        et("1 ^ 3", .{ .integer = 2 }),
        et("0 ^ 1", .{ .integer = 1 }),
        et(" true == false", .{ .boolean = false }),
        et("false == false", .{ .boolean = true }),
        et("false != false", .{ .boolean = false }),
        et(" true != false", .{ .boolean = true }),
        et(" true &&  true", .{ .boolean = true }),
        et(" true && false", .{ .boolean = false }),
        et(" true || false", .{ .boolean = true }),
        et("false || false", .{ .boolean = false }),
        et(" true  &  true", .{ .boolean = true }),
        et(" true  & false", .{ .boolean = false }),
        et(" true  | false", .{ .boolean = true }),
        et("false  | false", .{ .boolean = false }),
        et(" true  ^ false", .{ .boolean = true }),
        et("false  ^ false", .{ .boolean = false }),
        et("true   ^  true", .{ .boolean = false }),
        et("true", .{ .boolean = true }),
        et("false", .{ .boolean = false }),
        et("!true", .{ .boolean = false }),
        et("!false", .{ .boolean = true }),
        et(
            \\{
            \\  var a = 0
            \\  (a = 10) * a
            \\}
        , .{ .integer = 100 }),
        et(
            \\{
            \\  var a = 0
            \\  if a == 0
            \\    a = 10
            \\  a
            \\}
        , .{ .integer = 10 }),
        et(
            \\{
            \\  var a = 0
            \\  if a == 4
            \\    a = 10
            \\  a
            \\}
        , .{ .integer = 0 }),
        et(
            \\{
            \\  var a = 0
            \\  if a == 0
            \\    a = 10
            \\  else
            \\    a = 5
            \\  a
            \\}
        , .{ .integer = 10 }),
        et(
            \\{
            \\  var a = 0
            \\  if a == 4
            \\    a = 10
            \\  else
            \\    a = 5
            \\  a
            \\}
        , .{ .integer = 5 }),
        et(
            \\{
            \\  var i = 10
            \\  var result = 0
            \\  while i > 0
            \\  {
            \\    result = result + i
            \\    i = i - 1
            \\  }
            \\  result
            \\}
        , .{ .integer = 55 }),
        et(
            \\{
            \\  var result = 0
            \\  for i = 1 to 10
            \\    result = result + i
            \\  result
            \\}
        , .{ .integer = 55 }),
    }) |tt|
        t.runTest(
            state,
            EvaluatorTest,
            correctEvaluation,
            tt,
            std.fmt.allocPrint(t.allocator, "correct evaluation ({s} => {})", .{ tt.text, tt.expected_value }) catch unreachable,
            .allocated,
        );

    inline for (@typeInfo(one_shot_tests).@"struct".decls) |decl| {
        t.runTest(
            state,
            void,
            @field(one_shot_tests, decl.name),
            {},
            comptime camelToDisplay(decl.name),
            .static,
        );
    }
}
